File: src/z80.cpp
#include "z80.hpp"
#include "memory.hpp"
#include "port.hpp"

Z80::Z80(Memory* mem, Port* port) {
    // Store the memory and port pointers
    memory = mem;
    this->port = port;
    
    // Initialize main registers to zero
    AF = 0;
    BC = 0;
    DE = 0;
    HL = 0;
    
    // Initialize shadow registers to zero
    AF_ = 0;
    BC_ = 0;
    DE_ = 0;
    HL_ = 0;
    
    // Initialize index registers to zero
    IX = 0;
    IY = 0;
    
    // Initialize internal registers to zero
    SP = 0xFFFF;
    PC = 0;
    I = 0;
    R = 0;
    MEMPTR = 0;
    
    // Initialize interrupt-related registers and flags
    IFF1 = false;
    IFF2 = false;
    IM = 0;
    HALT = false;
    InterruptPending = false;
}

int Z80::ExecuteOneInstruction() {
    // Handle interrupts first if enabled
    if (IFF1 && InterruptPending) {
        InterruptPending = false;
        return HandleInterrupt();
    }
    InterruptPending = false;
    
    // Handle HALT state
    if (HALT) {
        return 4; // 4 T-states for HALT
    }

    // Read the first opcode byte
    uint8_t opcode = memory->ReadByte(PC);
    
    // Handle prefix opcodes
    switch (opcode) {
        case 0xDD: // DD prefix (IX instructions)
            // Increment PC past the prefix
            PC++;
            return ExecuteDDOpcode();
            
        case 0xFD: // FD prefix (IY instructions)
            // Increment PC past the prefix
            PC++;
            return ExecuteFDOpcode();
            
        case 0xCB: // CB prefix (bit manipulation instructions)
            // Increment PC past the prefix
            PC++;
            return ExecuteCBOpcode();
            
        case 0xED: // ED prefix (extended instructions)
            // Increment PC past the prefix
            PC++;
            return ExecuteEDOpcode();
            
        default: // Regular opcode
            return ExecuteOpcode();
    }
}

// HandleInterrupt handles interrupt processing
int Z80::HandleInterrupt() {
    // Exit HALT state
    // The HALT instruction halts the Z80; it does not increase the PC so that the instruction is re-
    // executed, until a maskable or non-maskable interrupt is accepted. Only then does the Z80 increase
    // the PC again and continues with the next instruction. During the HALT state, the HALT line is
    // set. The PC is increased before the interrupt routine is called.
    if (HALT) {
        HALT = false;
        PC++;  // Increment PC to exit HALT state
    }
    
    // Reset interrupt flip-flops
    IFF1 = false;
    IFF2 = false;
    
    // Handle interrupt based on mode
    switch (IM) {
        case 0:
        case 1:
            // Mode 0/1: Restart at address 0x0038
            Push(PC);
            PC = 0x0038;
            return 13; // 13 T-states for interrupt handling
            
        case 2:
            // Mode 2: Call interrupt vector
            Push(PC);
            {
                uint16_t vectorAddr = (uint16_t(I) << 8) | 0xFF; // Use 0xFF as vector for non-maskable interrupt
                PC = (uint16_t(memory->ReadByte(vectorAddr + 1)) << 8) | uint16_t(memory->ReadByte(vectorAddr));
            }
            return 19; // 19 T-states for interrupt handling
            
        default:
            // Should not happen, but handle gracefully
            return 0;
    }
}

// UpdateSZFlags updates the S and Z flags based on an 8-bit result
void Z80::UpdateSZFlags(uint8_t result) {
    SetFlag(FLAG_S, (result & 0x80) != 0);
    SetFlag(FLAG_Z, result == 0);
}

// UpdatePVFlags updates the P/V flag based on an 8-bit result (parity calculation)
void Z80::UpdatePVFlags(uint8_t result) {
    // Calculate parity (even number of 1-bits = 1, odd = 0)
    uint8_t parity = 1;
    for (int i = 0; i < 8; i++) {
        parity ^= (result >> i) & 1;
    }
    SetFlag(FLAG_PV, parity != 0);
}

// UpdateSZXYPVFlags updates the S, Z, X, Y, P/V flags based on an 8-bit result
void Z80::UpdateSZXYPVFlags(uint8_t result) {
    SetFlag(FLAG_S, (result & 0x80) != 0);
    SetFlag(FLAG_Z, result == 0);
    SetFlag(FLAG_X, (result & FLAG_X) != 0);
    SetFlag(FLAG_Y, (result & FLAG_Y) != 0);

    // Calculate parity (even number of 1-bits = 1, odd = 0)
    uint8_t parity = 1;
    for (int i = 0; i < 8; i++) {
        parity ^= (result >> i) & 1;
    }
    SetFlag(FLAG_PV, parity != 0);
}

// UpdateFlags3and5FromValue updates the X and Y flags from an 8-bit value
void Z80::UpdateFlags3and5FromValue(uint8_t value) {
    SetFlag(FLAG_X, (value & FLAG_X) != 0);
    SetFlag(FLAG_Y, (value & FLAG_Y) != 0);
}

// UpdateFlags3and5FromAddress updates the X and Y flags from the high byte of an address
void Z80::UpdateFlags3and5FromAddress(uint16_t address) {
    SetFlag(FLAG_X, ((address >> 8) & FLAG_X) != 0);
    SetFlag(FLAG_Y, ((address >> 8) & FLAG_Y) != 0);
}

// UpdateSZXYFlags updates the S, Z, X, Y flags based on an 8-bit result
void Z80::UpdateSZXYFlags(uint8_t result) {
    SetFlag(FLAG_S, (result & 0x80) != 0);
    SetFlag(FLAG_Z, result == 0);
    SetFlag(FLAG_X, (result & FLAG_X) != 0);
    SetFlag(FLAG_Y, (result & FLAG_Y) != 0);
}

// UpdateXYFlags updates the undocumented X and Y flags based on an 8-bit result
void Z80::UpdateXYFlags(uint8_t result) {
    SetFlag(FLAG_X, (result & FLAG_X) != 0);
    SetFlag(FLAG_Y, (result & FLAG_Y) != 0);
}

// GetFlag returns the state of a specific flag
bool Z80::GetFlag(uint8_t flag) {
    return (F & flag) != 0;
}

// SetFlag sets a flag to a specific state
void Z80::SetFlag(uint8_t flag, bool state) {
    if (state) {
        F |= flag;
    } else {
        F &= ~flag;
    }
}

// ClearFlag clears a specific flag
void Z80::ClearFlag(uint8_t flag) {
    F &= ~flag;
}

// ClearAllFlags clears all flags
void Z80::ClearAllFlags() {
    F = 0;
}


// ReadImmediateByte reads the next byte from memory at PC and increments PC
uint8_t Z80::ReadImmediateByte() {
    uint8_t value = memory->ReadByte(PC);
    PC++;
    return value;
}

// ReadImmediateWord reads the next word from memory at PC and increments PC by 2
uint16_t Z80::ReadImmediateWord() {
    uint8_t lo = memory->ReadByte(PC);
    PC++;
    uint8_t hi = memory->ReadByte(PC);
    PC++;
    return (uint16_t(hi) << 8) | uint16_t(lo);
}

// ReadDisplacement reads an 8-bit signed displacement value
int8_t Z80::ReadDisplacement() {
    int8_t value = int8_t(memory->ReadByte(PC));
    PC++;
    return value;
}

// ReadOpcode reads the next opcode from memory at PC and increments PC
uint8_t Z80::ReadOpcode() {
    uint8_t opcode = memory->ReadByte(PC);
    PC++;
    // Increment R register (memory refresh) for each opcode fetch
    // Note: R is a 7-bit register, bit 7 remains unchanged
    R = (R & 0x80) | ((R + 1) & 0x7F);
    return opcode;
}

// Push pushes a 16-bit value onto the stack
void Z80::Push(uint16_t value) {
    SP -= 2;
    memory->WriteByte(SP, uint8_t(value & 0xFF));
    memory->WriteByte(SP + 1, uint8_t((value >> 8) & 0xFF));
}

// Pop pops a 16-bit value from the stack
uint16_t Z80::Pop() {
    // Read low byte first, then high byte (little-endian)
    uint8_t lo = memory->ReadByte(SP);
    uint8_t hi = memory->ReadByte(SP + 1);
    SP += 2;
    return (uint16_t(hi) << 8) | uint16_t(lo);
}

// inc8 increments an 8-bit value and updates flags
uint8_t Z80::inc8(uint8_t value) {
    uint8_t result = value + 1;
    SetFlag(FLAG_H, (value & 0x0F) == 0x0F);
    SetFlag(FLAG_N, false);
    UpdateSZXYFlags(result);
    // Set PV flag if incrementing 0x7F to 0x80 (overflow from positive to negative)
    SetFlag(FLAG_PV, value == 0x7F);
    return result;
}

// dec8 decrements an 8-bit value and updates flags
uint8_t Z80::dec8(uint8_t value) {
    uint8_t result = value - 1;
    SetFlag(FLAG_H, (value & 0x0F) == 0x00);
    SetFlag(FLAG_N, true);
    UpdateSZXYFlags(result);
    // Set PV flag if decrementing 0x80 to 0x7F (overflow from negative to positive)
    SetFlag(FLAG_PV, value == 0x80);
    return result;
}

// rlca rotates the accumulator left circular
void Z80::rlca() {
    uint8_t result = (A << 1) | (A >> 7);
    A = result;
    SetFlag(FLAG_C, (A & 0x01) != 0);
    SetFlag(FLAG_H, false);
    SetFlag(FLAG_N, false);
    UpdateXYFlags(A);
}

// rla rotates the accumulator left through carry
void Z80::rla() {
    bool oldCarry = GetFlag(FLAG_C);
    uint8_t result = (A << 1);
    if (oldCarry) {
        result |= 0x01;
    }
    SetFlag(FLAG_C, (A & 0x80) != 0);
    A = result;
    SetFlag(FLAG_H, false);
    SetFlag(FLAG_N, false);
    UpdateXYFlags(A);
}

// rrca rotates the accumulator right circular
void Z80::rrca() {
    uint8_t result = (A >> 1) | (A << 7);
    A = result;
    SetFlag(FLAG_C, (A & 0x80) != 0);
    SetFlag(FLAG_H, false);
    SetFlag(FLAG_N, false);
    UpdateXYFlags(A);
}

// rra rotates the accumulator right through carry
void Z80::rra() {
    bool oldCarry = GetFlag(FLAG_C);
    uint8_t result = (A >> 1);
    if (oldCarry) {
        result |= 0x80;
    }
    SetFlag(FLAG_C, (A & 0x01) != 0);
    A = result;
    SetFlag(FLAG_H, false);
    SetFlag(FLAG_N, false);
    UpdateXYFlags(A);
}

// daa performs decimal adjust on accumulator
void Z80::daa() {
    uint8_t correction = 0;

    if (GetFlag(FLAG_H) || (A & 0x0F) > 9) {
        correction += 0x06;
    }

    if (A > 0x99 || GetFlag(FLAG_C)) {
        correction += 0x60;
        SetFlag(FLAG_C, true);
    }

    bool isSubtraction = GetFlag(FLAG_N);
    if (isSubtraction) {
        SetFlag(FLAG_H, GetFlag(FLAG_H) && (A & 0x0F) < 0x06);
        A -= correction;
    } else {
        SetFlag(FLAG_H, (A & 0x0F) > 9);
        A += correction;
    }

    SetFlag(FLAG_S, (A & 0x80) != 0);
    SetFlag(FLAG_Z, A == 0);
    SetFlag(FLAG_PV, parity(A));
    SetFlag(FLAG_X, (A & FLAG_X) != 0);
    SetFlag(FLAG_Y, (A & FLAG_Y) != 0);
}

// Helper function to calculate parity
bool Z80::parity(uint8_t val) {
    int count = 0;
    for (int i = 0; i < 8; i++) {
        if (val & (1 << i)) {
            count++;
        }
    }
    return (count % 2) == 0;
}

// cpl complements the accumulator
void Z80::cpl() {
    A = ~A;
    SetFlag(FLAG_H, true);
    SetFlag(FLAG_N, true);
    UpdateXYFlags(A);
}

// scf sets the carry flag
void Z80::scf() {
    SetFlag(FLAG_C, true);
    SetFlag(FLAG_N, false);
    SetFlag(FLAG_H, false);
    SetFlag(FLAG_X, (A & FLAG_X) != 0);
    SetFlag(FLAG_Y, (A & FLAG_Y) != 0);
}

// ccf complements the carry flag
void Z80::ccf() {
    bool oldCarry = GetFlag(FLAG_C);
    SetFlag(FLAG_C, !oldCarry);
    SetFlag(FLAG_H, oldCarry); // H = old C
    SetFlag(FLAG_N, false);
    SetFlag(FLAG_X, (A & FLAG_X) != 0);
    SetFlag(FLAG_Y, (A & FLAG_Y) != 0);
}

// add16 adds two 16-bit values and updates flags
uint16_t Z80::add16(uint16_t a, uint16_t b) {
    uint32_t result = (uint32_t)a + (uint32_t)b;
    SetFlag(FLAG_C, result > 0xFFFF);
    SetFlag(FLAG_H, (a & 0x0FFF) + (b & 0x0FFF) > 0x0FFF);
    SetFlag(FLAG_N, false);
    UpdateFlags3and5FromAddress((uint16_t)result);
    return (uint16_t)result;
}

// add8 adds an 8-bit value to the accumulator and updates flags
void Z80::add8(uint8_t value) {
    uint8_t a = A;
    uint16_t result = (uint16_t)a + (uint16_t)value;
    SetFlag(FLAG_C, result > 0xFF);
    SetFlag(FLAG_H, (a & 0x0F) + (value & 0x0F) > 0x0F);
    SetFlag(FLAG_N, false);
    UpdateSZXYFlags((uint8_t)result);
    // Set overflow flag: overflow occurs if adding two numbers with the same sign
    // produces a result with a different sign
    // Both operands have the same sign (both positive or both negative)
    // but the result has a different sign
    bool sameSign = ((a ^ value) & 0x80) == 0;
    bool differentResultSign = ((a ^ result) & 0x80) != 0;
    bool overflow = sameSign && differentResultSign;
    SetFlag(FLAG_PV, overflow);
    A = (uint8_t)result;
}

// adc8 adds an 8-bit value and carry to the accumulator and updates flags
void Z80::adc8(uint8_t value) {
    uint8_t a = A;
    bool carry = GetFlag(FLAG_C);
    uint16_t result;
    if (carry) {
        result = (uint16_t)a + (uint16_t)value + 1;
    } else {
        result = (uint16_t)a + (uint16_t)value;
    }
    SetFlag(FLAG_C, (int)a + (int)value + (int)(carry ? 1 : 0) > 0xFF);
    SetFlag(FLAG_H, (a & 0x0F) + (value & 0x0F) + (carry ? 1 : 0) > 0x0F);
    SetFlag(FLAG_N, false);
    UpdateSZXYFlags((uint8_t)result);
    // Set overflow flag: overflow occurs if adding two numbers with the same sign
    // produces a result with a different sign
    uint8_t originalValue = a;
    if (carry) {
        bool sameSign = ((originalValue ^ value) & 0x80) == 0;
        bool differentResultSign = ((originalValue ^ result) & 0x80) != 0;
        bool overflow = sameSign && differentResultSign;
        SetFlag(FLAG_PV, overflow);
    } else {
        bool sameSign = ((originalValue ^ value) & 0x80) == 0;
        bool differentResultSign = ((originalValue ^ result) & 0x80) != 0;
        bool overflow = sameSign && differentResultSign;
        SetFlag(FLAG_PV, overflow);
    }
    A = (uint8_t)result;
}

// sub8 subtracts an 8-bit value from the accumulator and updates flags
void Z80::sub8(uint8_t value) {
    uint8_t a = A;
    uint16_t result = (uint16_t)a - (uint16_t)value;
    SetFlag(FLAG_C, a < value);
    SetFlag(FLAG_H, (a & 0x0F) < (value & 0x0F));
    SetFlag(FLAG_N, true);
    UpdateSZXYFlags((uint8_t)result);
    // Set overflow flag: overflow occurs if subtracting two numbers with different signs
    // produces a result with the same sign as the subtrahend
    bool overflow = (((a ^ value) & 0x80) != 0) && (((a ^ result) & 0x80) != 0);
    SetFlag(FLAG_PV, overflow);
    A = (uint8_t)result;
}

// sbc8 subtracts an 8-bit value and carry from the accumulator and updates flags
void Z80::sbc8(uint8_t value) {
    uint8_t a = A;
    bool carry = GetFlag(FLAG_C);
    uint16_t result;
    if (carry) {
        result = (uint16_t)a - (uint16_t)value - 1;
    } else {
        result = (uint16_t)a - (uint16_t)value;
    }
    SetFlag(FLAG_C, (int)a - (int)value - (int)(carry ? 1 : 0) < 0);
    SetFlag(FLAG_H, (a & 0x0F) < (value & 0x0F) + (carry ? 1 : 0));
    SetFlag(FLAG_N, true);
    UpdateSZXYFlags((uint8_t)result);
    // Set overflow flag: overflow occurs if subtracting two numbers with different signs
    // produces a result with the same sign as the subtrahend
    uint8_t originalValue = a;
    if (carry) {
        bool overflow = (((originalValue ^ value) & 0x80) != 0) && (((originalValue ^ result) & 0x80) != 0);
        SetFlag(FLAG_PV, overflow);
    } else {
        bool overflow = (((originalValue ^ value) & 0x80) != 0) && (((originalValue ^ result) & 0x80) != 0);
        SetFlag(FLAG_PV, overflow);
    }
    A = (uint8_t)result;
}

// and8 performs bitwise AND with the accumulator and updates flags
void Z80::and8(uint8_t value) {
    A &= value;
    SetFlag(FLAG_C, false);
    SetFlag(FLAG_N, false);
    SetFlag(FLAG_H, true);
    UpdateSZXYFlags(A);
    // For logical operations, P/V flag indicates parity
    SetFlag(FLAG_PV, parity(A));
}

// xor8 performs bitwise XOR with the accumulator and updates flags
void Z80::xor8(uint8_t value) {
    A ^= value;
    SetFlag(FLAG_C, false);
    SetFlag(FLAG_N, false);
    SetFlag(FLAG_H, false);
    UpdateSZXYFlags(A);
    // For logical operations, P/V flag indicates parity
    SetFlag(FLAG_PV, parity(A));
}

// or8 performs bitwise OR with the accumulator and updates flags
void Z80::or8(uint8_t value) {
    A |= value;
    SetFlag(FLAG_C, false);
    SetFlag(FLAG_N, false);
    SetFlag(FLAG_H, false);
    UpdateSZXYFlags(A);
    // For logical operations, P/V flag indicates parity
    SetFlag(FLAG_PV, parity(A));
}

// cp8 compares an 8-bit value with the accumulator and updates flags
void Z80::cp8(uint8_t value) {
    uint8_t a = A;
    uint16_t result = (uint16_t)a - (uint16_t)value;
    SetFlag(FLAG_C, a < value);
    SetFlag(FLAG_H, (a & 0x0F) < (value & 0x0F));
    SetFlag(FLAG_N, true);
    UpdateSZFlags((uint8_t)result);
    // For CP instruction, X and Y flags are set from the operand, not the result
    UpdateFlags3and5FromValue(value);
    // Set overflow flag: overflow occurs if subtracting two numbers with different signs
    // produces a result with the same sign as the subtrahend
    bool overflow = (((a ^ value) & 0x80) != 0) && (((a ^ result) & 0x80) != 0);
    SetFlag(FLAG_PV, overflow);
}

File: src/z80.go
// Package z80 implements a Z80 CPU emulator with support for all documented
// and undocumented opcodes, flags, and registers.
package z80


// Memory interface for memory operations
type Memory interface {
	ReadByte(address uint16) byte
	WriteByte(address uint16, value byte)
	ReadWord(address uint16) uint16
	WriteWord(address uint16, value uint16)
}

// IO interface for input/output operations
type IO interface {
	ReadPort(port uint16) byte
	WritePort(port uint16, value byte)
	CheckInterrupt() bool
}

// FLAG_* constants represent the bit positions of the FLAGS register
const (
	FLAG_C  = 0x01 // Carry flag
	FLAG_N  = 0x02 // Add/Subtract flag
	FLAG_PV = 0x04 // Parity/Overflow flag
	FLAG_3  = 0x08 // Undocumented 3rd bit flag
	FLAG_H  = 0x10 // Half Carry flag
	FLAG_5  = 0x20 // Undocumented 5th bit flag
	FLAG_Z  = 0x40 // Zero flag
	FLAG_S  = 0x80 // Sign flag
)

// FLAG_X and FLAG_Y are aliases for FLAG_3 and FLAG_5 respectively
const (
	FLAG_X = FLAG_3
	FLAG_Y = FLAG_5
)

// CPU represents the state of a Z80 processor
type CPU struct {
	A      byte   // Accumulator
	F      byte   // Flags register
	B, C   byte   // BC register pair
	D, E   byte   // DE register pair
	H, L   byte   // HL register pair
	A_, F_ byte   // Alternate AF register pair
	B_, C_ byte   // Alternate BC register pair
	D_, E_ byte   // Alternate DE register pair
	H_, L_ byte   // Alternate HL register pair
	IX     uint16 // Index register X
	IY     uint16 // Index register Y
	SP     uint16 // Stack pointer
	PC     uint16 // Program counter
	I      byte   // Interrupt vector
	R      byte   // Memory refresh
	IM     byte   // Interrupt mode (0, 1, or 2)
	IFF1   bool   // Interrupt flip-flop 1
	IFF2   bool   // Interrupt flip-flop 2
	HALT   bool   // HALT state flag
	MEMPTR uint16 // MEMPTR register (undocumented)

	Memory Memory // Memory interface
	IO     IO     // IO interface
}

// New creates a new Z80 CPU instance
func New(memory Memory, io IO) *CPU {
	return &CPU{
		Memory: memory,
		IO:     io,
	}
}

// GetBC returns the combined value of the B and C registers
func (cpu *CPU) GetBC() uint16 {
	return (uint16(cpu.B) << 8) | uint16(cpu.C)
}

// GetDE returns the combined value of the D and E registers
func (cpu *CPU) GetDE() uint16 {
	return (uint16(cpu.D) << 8) | uint16(cpu.E)
}

// GetHL returns the combined value of the H and L registers
func (cpu *CPU) GetHL() uint16 {
	return (uint16(cpu.H) << 8) | uint16(cpu.L)
}

// SetBC sets the B and C registers from a 16-bit value
func (cpu *CPU) SetBC(value uint16) {
	cpu.B = byte(value >> 8)
	cpu.C = byte(value)
}

// SetDE sets the D and E registers from a 16-bit value
func (cpu *CPU) SetDE(value uint16) {
	cpu.D = byte(value >> 8)
	cpu.E = byte(value)
}

// SetHL sets the H and L registers from a 16-bit value
func (cpu *CPU) SetHL(value uint16) {
	cpu.H = byte(value >> 8)
	cpu.L = byte(value)
}

// GetIXH returns the high byte of the IX register
func (cpu *CPU) GetIXH() byte {
	return byte(cpu.IX >> 8)
}

// GetIXL returns the low byte of the IX register
func (cpu *CPU) GetIXL() byte {
	return byte(cpu.IX)
}

// GetIYH returns the high byte of the IY register
func (cpu *CPU) GetIYH() byte {
	return byte(cpu.IY >> 8)
}

// GetIYL returns the low byte of the IY register
func (cpu *CPU) GetIYL() byte {
	return byte(cpu.IY)
}

// SetIXH sets the high byte of the IX register
func (cpu *CPU) SetIXH(value byte) {
	cpu.IX = (cpu.IX & 0x00FF) | (uint16(value) << 8)
}

// SetIXL sets the low byte of the IX register
func (cpu *CPU) SetIXL(value byte) {
	cpu.IX = (cpu.IX & 0xFF00) | uint16(value)
}

// SetIYH sets the high byte of the IY register
func (cpu *CPU) SetIYH(value byte) {
	cpu.IY = (cpu.IY & 0x00FF) | (uint16(value) << 8)
}

// SetIYL sets the low byte of the IY register
func (cpu *CPU) SetIYL(value byte) {
	cpu.IY = (cpu.IY & 0xFF00) | uint16(value)
}

// GetFlag returns the state of a specific flag
func (cpu *CPU) GetFlag(flag byte) bool {
	return (cpu.F & flag) != 0
}

// SetFlag sets a flag to a specific state
func (cpu *CPU) SetFlag(flag byte, state bool) {
	if state {
		cpu.F |= flag
	} else {
		cpu.F &^= flag
	}
}

// SetFlagState is a helper function to set a flag based on a boolean condition
func (cpu *CPU) SetFlagState(flag byte, condition bool) {
	if condition {
		cpu.F |= flag
	} else {
		cpu.F &^= flag
	}
}

// ClearFlag clears a specific flag
func (cpu *CPU) ClearFlag(flag byte) {
	cpu.F &^= flag
}

// ClearAllFlags clears all flags
func (cpu *CPU) ClearAllFlags() {
	cpu.F = 0
}

// ReadImmediateByte reads the next byte from memory at PC and increments PC
func (cpu *CPU) ReadImmediateByte() byte {
	value := cpu.Memory.ReadByte(cpu.PC)
	cpu.PC++
	return value
}

// ReadImmediateWord reads the next word from memory at PC and increments PC by 2
func (cpu *CPU) ReadImmediateWord() uint16 {
	lo := cpu.Memory.ReadByte(cpu.PC)
	cpu.PC++
	hi := cpu.Memory.ReadByte(cpu.PC)
	cpu.PC++
	return (uint16(hi) << 8) | uint16(lo)
}

// ReadDisplacement reads an 8-bit signed displacement value
func (cpu *CPU) ReadDisplacement() int8 {
	value := int8(cpu.Memory.ReadByte(cpu.PC))
	cpu.PC++
	return value
}

// ReadOpcode reads the next opcode from memory at PC and increments PC
func (cpu *CPU) ReadOpcode() byte {
	opcode := cpu.Memory.ReadByte(cpu.PC)
	cpu.PC++
	// Increment R register (memory refresh) for each opcode fetch
	// Note: R is a 7-bit register, bit 7 remains unchanged
	cpu.R = (cpu.R & 0x80) | ((cpu.R + 1) & 0x7F)
	return opcode
}

// Push pushes a 16-bit value onto the stack
func (cpu *CPU) Push(value uint16) {
	cpu.SP -= 2
	cpu.Memory.WriteWord(cpu.SP, value)
}

// Pop pops a 16-bit value from the stack
func (cpu *CPU) Pop() uint16 {
	// Read low byte first, then high byte (little-endian)
	lo := cpu.Memory.ReadByte(cpu.SP)
	hi := cpu.Memory.ReadByte(cpu.SP + 1)
	cpu.SP += 2
	return (uint16(hi) << 8) | uint16(lo)
}

// UpdateSZFlags updates the S and Z flags based on an 8-bit result
func (cpu *CPU) UpdateSZFlags(result byte) {
	cpu.SetFlagState(FLAG_S, (result&0x80) != 0)
	cpu.SetFlagState(FLAG_Z, result == 0)
}

func (cpu *CPU) UpdatePVFlags(result byte) {
	// Calculate parity (even number of 1-bits = 1, odd = 0)
	parity := byte(1)
	for i := 0; i < 8; i++ {
		parity ^= (result >> i) & 1
	}
	cpu.SetFlagState(FLAG_PV, parity != 0)
}

// UpdateSZXYPVFlags updates the S, Z, X, Y, P/V flags based on an 8-bit result
func (cpu *CPU) UpdateSZXYPVFlags(result byte) {
	cpu.SetFlagState(FLAG_S, (result&0x80) != 0)
	cpu.SetFlagState(FLAG_Z, result == 0)
	cpu.SetFlagState(FLAG_X, (result&FLAG_X) != 0)
	cpu.SetFlagState(FLAG_Y, (result&FLAG_Y) != 0)

	// Calculate parity (even number of 1-bits = 1, odd = 0)
	parity := byte(1)
	for i := 0; i < 8; i++ {
		parity ^= (result >> i) & 1
	}
	cpu.SetFlagState(FLAG_PV, parity != 0)
}

// UpdateFlags3and5FromValue updates the X and Y flags from an 8-bit value
func (cpu *CPU) UpdateFlags3and5FromValue(value byte) {
	cpu.SetFlagState(FLAG_X, (value&FLAG_X) != 0)
	cpu.SetFlagState(FLAG_Y, (value&FLAG_Y) != 0)
}

// UpdateFlags3and5FromAddress updates the X and Y flags from the high byte of an address
func (cpu *CPU) UpdateFlags3and5FromAddress(address uint16) {
	cpu.SetFlagState(FLAG_X, (byte(address>>8)&FLAG_X) != 0)
	cpu.SetFlagState(FLAG_Y, (byte(address>>8)&FLAG_Y) != 0)
}

// UpdateSZXYFlags updates the S, Z, X, Y flags based on an 8-bit result
func (cpu *CPU) UpdateSZXYFlags(result byte) {
	cpu.SetFlagState(FLAG_S, (result&0x80) != 0)
	cpu.SetFlagState(FLAG_Z, result == 0)
	cpu.SetFlagState(FLAG_X, (result&FLAG_X) != 0)
	cpu.SetFlagState(FLAG_Y, (result&FLAG_Y) != 0)
}

// boolToByte converts a boolean to a byte (1 if true, 0 if false)
func boolToByte(b bool) byte {
	if b {
		return 1
	}
	return 0
}

// ExecuteOneInstruction executes a single instruction and returns the number of T-states used
func (cpu *CPU) ExecuteOneInstruction() int {
	// Handle interrupts first if enabled
	if cpu.IFF1 && cpu.IO.CheckInterrupt() {
		return cpu.HandleInterrupt()
	}
	//fmt.Printf("%b",cpu.HALT)
	// Handle HALT state
	if cpu.HALT {
		return 4 // 4 T-states for HALT
	}

	// Read the next opcode
	opcode := cpu.ReadOpcode()

	// Handle prefixed opcodes
	switch opcode {
	case 0xCB:
		cbOpcode := cpu.ReadOpcode()
		return cpu.ExecuteCBOpcode(cbOpcode)
	case 0xDD:
		ddOpcode := cpu.ReadOpcode()
		return cpu.ExecuteDDOpcode(ddOpcode)
	case 0xED:
		edOpcode := cpu.ReadOpcode()
		return cpu.ExecuteEDOpcode(edOpcode)
	case 0xFD:
		fdOpcode := cpu.ReadOpcode()
		return cpu.ExecuteFDOpcode(fdOpcode)
	default:
		return cpu.ExecuteOpcode(opcode)
	}
}

// HandleInterrupt handles interrupt processing
func (cpu *CPU) HandleInterrupt() int {
	// Exit HALT state
	// The HALT instruction halts the Z80; it does not increase the PC so that the instruction is re-
	//executed, until a maskable or non-maskable interrupt is accepted. Only then does the Z80 increase
	//the PC again and continues with the next instruction. During the HALT state, the HALT line is
	//set. The PC is increased before the interrupt routine is called.
	if(cpu.HALT) {
		cpu.HALT = false
		cpu.PC++
	}
	
	
	// Reset interrupt flip-flops
	cpu.IFF1 = false
	cpu.IFF2 = false
	// Handle interrupt based on mode
	switch cpu.IM {
	case 0, 1:
		// Mode 0/1: Restart at address 0x0038
		cpu.Push(cpu.PC)
		cpu.PC = 0x0038
		return 13 // 13 T-states for interrupt handling
	case 2:
		// Mode 2: Call interrupt vector
		cpu.Push(cpu.PC)
		vectorAddr := (uint16(cpu.I) << 8) | 0xFF // Use 0xFF as vector for non-maskable interrupt
		cpu.PC = cpu.Memory.ReadWord(vectorAddr)
		return 19 // 19 T-states for interrupt handling
	default:
		// Should not happen, but handle gracefully
		return 0
	}
}

// HandleNMI handles non-maskable interrupt
func (cpu *CPU) HandleNMI() int {
	// Save current PC on stack
	cpu.Push(cpu.PC)

	// Jump to NMI handler
	cpu.PC = 0x0066

	// Disable interrupts
	cpu.IFF1 = false

	return 11 // 11 T-states for NMI handling
}

// GetAF returns the combined value of the A and F registers
func (cpu *CPU) GetAF() uint16 {
	return (uint16(cpu.A) << 8) | uint16(cpu.F)
}

// SetAF sets the A and F registers from a 16-bit value
func (cpu *CPU) SetAF(value uint16) {
	cpu.A = byte(value >> 8)
	cpu.F = byte(value)
}

// GetAF_ returns the combined value of the alternate A_ and F_ registers
func (cpu *CPU) GetAF_() uint16 {
	return (uint16(cpu.A_) << 8) | uint16(cpu.F_)
}

// SetAF_ sets the alternate A_ and F_ registers from a 16-bit value
func (cpu *CPU) SetAF_(value uint16) {
	cpu.A_ = byte(value >> 8)
	cpu.F_ = byte(value)
}

// GetBC_ returns the combined value of the alternate B_ and C_ registers
func (cpu *CPU) GetBC_() uint16 {
	return (uint16(cpu.B_) << 8) | uint16(cpu.C_)
}

// GetDE_ returns the combined value of the alternate D_ and E_ registers
func (cpu *CPU) GetDE_() uint16 {
	return (uint16(cpu.D_) << 8) | uint16(cpu.E_)
}

// GetHL_ returns the combined value of the alternate H_ and L_ registers
func (cpu *CPU) GetHL_() uint16 {
	return (uint16(cpu.H_) << 8) | uint16(cpu.L_)
}

// SetBC_ sets the alternate B_ and C_ registers from a 16-bit value
func (cpu *CPU) SetBC_(value uint16) {
	cpu.B_ = byte(value >> 8)
	cpu.C_ = byte(value)
}

// SetDE_ sets the alternate D_ and E_ registers from a 16-bit value
func (cpu *CPU) SetDE_(value uint16) {
	cpu.D_ = byte(value >> 8)
	cpu.E_ = byte(value)
}

// SetHL_ sets the alternate H_ and L_ registers from a 16-bit value
func (cpu *CPU) SetHL_(value uint16) {
	cpu.H_ = byte(value >> 8)
	cpu.L_ = byte(value)
}

// UpdateXYFlags updates the undocumented X and Y flags based on an 8-bit result
func (cpu *CPU) UpdateXYFlags(result byte) {
	cpu.SetFlagState(FLAG_X, (result&FLAG_X) != 0)
	cpu.SetFlagState(FLAG_Y, (result&FLAG_Y) != 0)
}
